### DISCUSSION TITLE
[Curso 2811] Dúvida com ViewModel

### DISCUSSION THREAD

**Discussion by juliusczm** [balta-io/forum#680] [Tira Dúvidas] [answered]
Estou implementando os recursos adicionais para a API, conforme  #678 .

Estou tendo que adicionar ViewModels para que seja feita a compatibilização dos recursos desejados, uma vez que em raros cenários o próprio modelo se adequa a todas as atividades do CRUD.

Nos cenários em que relacionamentos entre os modelos, estou com um pouco de dificuldade de saber como modelar as ViewModels de uma forma mais inteligente.

A título de exemplo, considere os modelos de User e Role.
```csharp
public class User
  {
      public int Id { get; set; }
      public string Name { get; set; }
      public string Email { get; set; }
      public string PasswordHash { get; set; }
      public string Image { get; set; }
      public string Slug { get; set; }
      public string Bio { get; set; }

      public IList<Post> Posts { get; set; }
      public IList<Role> Roles { get; set; }
  }
```

```csharp
  public class Role
  {
      public int Id { get; set; }
      public string Name { get; set; }
      public string Slug { get; set; }

      public IList<User> Users { get; set; }
  }
```

Comecei modelando o CRUD de Role com o Get básico. Para isso, criei uma RoleViewModel  que, a principio, atenderia esse requisito.
```csharp
public class RoleViewModel
{
    public int Id { get; set; }
    public string Name { get; set; }
    public string Slug { get; set; }
}
```

Agora, se eu quiser trazer os usuários associados a cada um dos perfis, imagino que deva ser necessária outra ViewModel. Contudo, se eu não quiser trazer todas as propriedades dos Usuários, antes da ViewModel eu teria que criar um outro modelo pro User ? É esse o raciocínio?

```csharp
public class UserSummaryViewModel
{
    public int Id { get; set; }
    public string Name { get; set; }
    public string Email { get; set; }
}
```

```csharp
public class ListRoleUserViewModel
{
    public int Id { get; set; }
    public string Name { get; set; }
    public string Slug { get; set; }

    public List<UserSummaryViewModel> Users { get; set; }
}
```

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @juliusczm, como vai?

Seguindo ao pé da letra, sim, teria que ser outra ViewModel, porém, se precisa dos dados de uma entidade (Todos), não vejo problema em trazer a entidade em si na ViewModel...

Porém, no caso de usuário, muda um pouco pois os dados são mais sensíveis... talvez o melhor caminho seja mesmo criar as ViewModels!
